### ISSUE TITLE
nvptx abi adjustment may keep PassMode::Direct for small aggregates in conv `extern "C"`

### ISSUE THREAD

**Issue by kjetilkjeka** [rust-lang/rust#117480]
@kjetilkjeka I had to disable this test on nvptx64 since the `extern "C"` ABI uses `Direct` pass mode in invalid ways. I think that's caused by this code:

https://github.com/rust-lang/rust/blob/b75b3b3afecfb6638c7c75ad665f8f3091aab1ae/compiler/rustc_target/src/abi/call/nvptx64.rs#L10-L14

That's invalid since it doesn't say what to do for smaller aggregates, and they default to the (bad) `Direct`. Instead you have to say which register they are supposed to be passed in. You can check what the other targets are doing. Targets are expected to set an explicit pass mode for *all* aggregate arguments.

This should be easy to reproduce by having a function like
```rust
#[repr(C)]
struct ReprC1<T: ?Sized>(T);

extern "C" fn myfn(x: ReprC1<i32>) {}
```
which will likely ICE on the current compiler already.

_Originally posted by @RalfJung in https://github.com/rust-lang/rust/pull/117351#discussion_r1375418626_

**Comment by kjetilkjeka** [comment]
Thanks for pointing this out @RalfJung 

As a note I had to use `#[no_mangle]` to avoid the compiler from removing the symbol before the assert was hit. 

Do you know why this test started failing with your changes in #117351 and not before that?

**Comment by RalfJung** [comment]
> Do you know why this test started failing with your changes in https://github.com/rust-lang/rust/pull/117351 and not before that?

We used to only run the assertion in codegen, i.e. when a function with such argument types actually gets generated or called. The abi compatibility test works on function pointers, it computes their ABI but never calls them so we never enter the part of the codegen backend that had the assertion.

I'm going to try to find a way to move this check to the ABI computation logic, but that turns out to be hard since we do compute some ABIs that we don't actually support...

**Comment by RalfJung** [comment]
> I'm curious to how to make the tests catch it if we have a regression and start producing PassMode::Direct after this issue is resolved

https://github.com/rust-lang/rust/pull/117500 does ensure we'd catch this.

**Comment by RalfJung** [comment]
This should not have been closed without enabling the relevant tests for nvtpx:
https://github.com/rust-lang/rust/pull/125694.

**Comment by RalfJung** [comment]
Actually the issue is still open -- the test fails:
```
`PassMode::Direct` for aggregates only allowed for "unadjusted" and "ptx-kernel" functions and on wasm
Problematic type: TyAndLayout {
    ty: ReprC1<*const i32>,
    layout: Layout {
        size: Size(8 bytes),
        align: AbiAndPrefAlign {
            abi: Align(8 bytes),
            pref: Align(8 bytes),
        },
        abi: Aggregate {
            sized: true,
        },
        fields: Arbitrary {
            offsets: [
                Size(0 bytes),
            ],
            memory_index: [
                0,
            ],
        },
        largest_niche: None,
        variants: Single {
            index: 0,
        },
        max_repr_align: None,
        unadjusted_abi_align: Align(8 bytes),
    },
}
```

**Comment by RalfJung** [comment]
https://github.com/rust-lang/rust/pull/117671 only changed `classify_arg`, shouldn't it change `classify_ret` as well?

**Comment by RalfJung** [comment]
> `.param .align 1 .b8 <name>[1]`

So if this was a struct with an i32 field it would be `[4]` at the end to indicate an array of length 4?

> When I try to cast to a single 8-bit register it passes it as a u8 instead of the SingleU8. That is, as you say, due to it passing it as the underlying type of single-element arrays.

How are you setting up the cast?
When the `rest` is a u8 then it will indeed become a single u8. But when one `prefix` element is a `u8`, that should become a struct for LLVM, which is exactly the same as with `Direct`.

**Comment by kjetilkjeka** [comment]
> So if this was a struct with an i32 field it would be [4] at the end to indicate an array of length 4?

Yes, this is correct.

> How are you setting up the cast?
When the rest is a u8 then it will indeed become a single u8. But when one prefix element is a u8, that should become a struct for LLVM, which is exactly the same as with Direct.

Ahh, that is not intuitive at all for me, but it works indeed. I made the return tests pass with this code.

```rust
   if ret.layout.is_aggregate() {
        let align_bytes = ret.layout.align.abi.bytes();
        let size = ret.layout.size;

        let reg = match align_bytes {
            1 => Reg::i8(),
            2 => Reg::i16(),
            4 => Reg::i32(),
            8 => Reg::i64(),
            16 => Reg::i128(),
            _ => unreachable!("Align is given as power of 2 no larger than 16 bytes"),
        };

        if align_bytes == size.bytes() {
            ret.cast_to(CastTarget {
                prefix: [Some(reg), None, None, None, None, None, None, None],
                rest: Uniform::new(Reg::i8(), Size::from_bytes(0)),
                attrs: ArgAttributes {
                    regular: ArgAttribute::default(),
                    arg_ext: ArgExtension::None,
                    pointee_size: Size::ZERO,
                    pointee_align: None,
                },
            });
        } else {
            ret.cast_to(Uniform::new(reg, size));
        }
    } else if ret.layout.size.bits() < 32 {
        ret.extend_integer_width_to(32);
    }
```

I will most likely get a chance to create a PR on Friday. Thanks for that trick.

**Comment by RalfJung** [comment]
Yeah to be clear I have no idea how CastTarget is intended to be used, I just read the code for how it gets translated into an LLVM type, and then found a way to make that generate the LLVM type you want, for the one case you mentioned. I have no idea how to turn that into a general principle that works for all types.

You seem to make a special case for `align_bytes == size.bytes()`, but doesn't that mean that for a type with two `i32` fields we'll get the wrong encoding? Specifically, currently that will be `{ i32, i32 }` (a struct with two i32 fields), with your patch that will be `[2 x i32]` (a two-element array of `i32`). Maybe those generate the same PTX output, maybe not -- please carefully check that.

**Comment by kjetilkjeka** [comment]
> You seem to make a special case for align_bytes == size.bytes(), but doesn't that mean that for a type with two i32 fields we'll get the wrong encoding? Specifically, currently that will be { i32, i32 } (a struct with two i32 fields), with your patch that will be [2 x i32] (a two-element array of i32). Maybe those generate the same PTX output, maybe not -- please carefully check that.

I have specifically added a test for this type in #125980. It seems to produce correct results.

**Comment by bjorn3** [comment]
Fixed by https://github.com/rust-lang/rust/pull/125980 and https://github.com/rust-lang/rust/pull/133932
